### ISSUE TITLE
Git History Support

### ISSUE THREAD

**Issue by regicsolutions** [cookiecutter/cookiecutter#1503]
* Cookiecutter version: latest
* Template project url: NA
* Python version: 3.8.6
* Operating System: Win10

### Description:

Looking to see if cookiecutter offers a way to easily retain git history to manage template updates. I could not find any documentation on this topic.

### What I've run:

```
**The issue i am trying to solve:** I just created 50 micro-services using one template, a month later i need to make a change to the template, lets say it was a new Jenkinsfile version, there is no way for me to be able to retrieve that update in my post-generated repo.
```

**Comment by HappyEinara** [comment]
I am working on a little package that extends cookiecutter, which I hope to release soon for my own use but also to demonstrate a similar suggestion to yours to the cookiecutter devs. Here are some thoughts:

- Cookiecutter already injects `_template` and `_output_dir` metadata fields into the context.
- There are other metadata items I'd like available to my templates and especially my post-gen hooks
- In particular I'd like more detail on the template repo: the hash and tags of the specific commit that is used to generate my project, but this could extend to other insights (author or date of the template version, the commit message at the template HEAD...)
- Some bits of git/template info which we already know (such as the branch we want checking out) can currently be passed in using `extra_context`. Injecting `extra_context` on the command line is not documented but [it's there](https://github.com/cookiecutter/cookiecutter/blob/52dd18513bbab7f0fbfcb2938c9644d9092247cf/cookiecutter/cli.py#L73). So for example, you could do `cookiecutter gh:my_repo --checkout my_branch template_git_branch=my_branch`. However that option only extends to things you know when you issue the command and it's a bit of clunky way to do it.
- There's no reason why the automatically injected metadata couldn't be richer, but perhaps it could be kept out of the `cookiecutter` key on the context to avoid confusion or clashes. My suggestion:
```
{
  "cookiecutter": {
    "project_name": "happys_project",
    "project_slug": "happys-project"
  },
  "_meta": {
    "template": "gh:HappyEinara/happys-template",
    "template_location": "~/.cookiecutters/happys-template",
    "template_git_commit": "abc123d",
    "template_git_branch": "develop",
    "template_directory": "/lambda_api",
    "template_git_tags": ["v0.1.23-develop"],
  }
}
```
etc.
- A post-gen hook could then do `{{ _meta.template }}` instead of `{{ cookicutter._template }}`, but also `{{ _meta.template_git_commit }}` for example.
- An alternative idea: pre-/post-gen hooks are great to have but they execute outside of cookiecutter and only have the populated context available. Perhaps a plugin mechanism could let us register a callback, maybe [somewhere around here](https://github.com/cookiecutter/cookiecutter/blob/52dd18513bbab7f0fbfcb2938c9644d9092247cf/cookiecutter/main.py#L105), that would let us inspect the current state of the input repo, the context, the envvars, the extra_context, etc exactly as it is all available to cookiecutter, and update the context dict in python as we choose. Something like `run_callbacks("pre-generate", context, repo, env_vars)`. That would add complexity to an otherwise simple tool, however and most use cases could be solved just by making the auto-injected metadata richer.
- Another solution is for the script that calls cookiecutter to pre-clone the template, inspect it, and add extra_context to the cookiecutter command. That's also clunky, requires a second clone, and there's a race condition.
- Yet another solution is for a post-gen hook to try to inspect the cloned repo in `~/.cookiecutters`, trusting cookiecutter to have left it in the state used for template generation, but that's also not a robust solution.

The project I mentioned I was working on monkeypatches cookiecutter to add a `_meta` dict as described above and I'll let you see it when it's tested. Monkeypatching is not the answer, but I want to demonstrate an example of the hooks extra metadata would help me achieve, and if the core devs reject this request, it'll do the job.

In summary my preferred solution is that the existing `_`-prefixed values be moved to a subdict called `_meta` as a sibling of `cookiecutter` and be enriched with all the extra context we could need but can't currently get easily or robustly. The community (and OP) could feed back on the values that might be added there. A separate meta dict also permits future expansion better than the current approach of adding one or two `_` keys on the main context.

Hope this helps stimulate thought!

**Comment by HappyEinara** [comment]
(Another thought on implementation if something along these lines is considered: cc currently uses subprocess to call `git` and `hg` commands. VCS-specific code could be factored out to a generic repo-handling module that itself depends on encapsulated service classes/modules for git and Mercurial, and use proper python libraries to interact with each. It's unlikely to be much in demand, but other VCSes could easily be implemented to provide the interface cc wants by adding a new module for that VCS.)

**Comment by samj1912** [comment]
https://github.com/cruft/cruft already handles this and is fully compatible with cookiecutter templates :)

**Comment by HappyEinara** [comment]
> https://github.com/cruft/cruft already handles this and is fully compatible with cookiecutter templates :) 

@samj1912 THANK YOU FOR SHARING!

I've only taken a cursory look so far but it looks a lot like the sort of thing I've been hoping to achieve myself. It's been worthwhile anyway because I've learned a lot about python imports as I've tried to implement this with monkeypatches!

You may have just opened an awesome door though, and I'm very grateful. I'll give it a go and feed back over there. Thanks again mate!
